zebra-state/src/sled_state.rs

```diff
@@ -0,0 +1,216 @@
+/// clones of handles for [`sled::Tree`]s) into the futures they return.
+///
+/// This means that the returned futures have a `'static` lifetime and don't
+/// borrow any resources from the [`SledState`], and the actual database work is
+/// performed asynchronously when the returned future is polled, not while it is
+/// created.  This is analogous to the way [`tower::Service::call`] works.
+pub struct SledState {
+    /// Queued blocks that arrived out of order, indexed by their parent block hash.
+    queued_by_prev_hash: HashMap<block::Hash, QueuedBlock>,
+
+    hash_by_height: sled::Tree,
+    height_by_hash: sled::Tree,
+    block_by_height: sled::Tree,
+    // tx_by_hash: sled::Tree,
+    // utxo_by_outpoint: sled::Tree,
+    // sprout_nullifiers: sled::Tree,
+    // sapling_nullifiers: sled::Tree,
+    // sprout_anchors: sled::Tree,
+    // sapling_anchors: sled::Tree,
+}
+
+impl SledState {
+    pub fn new(config: &Config, network: Network) -> Self {
+        let db = config.sled_config(network).open().unwrap();
+
+        Self {
+            queued_by_prev_hash: HashMap::new(),
+            hash_by_height: db.open_tree(b"hash_by_height").unwrap(),
+            height_by_hash: db.open_tree(b"height_by_hash").unwrap(),
+            block_by_height: db.open_tree(b"block_by_height").unwrap(),
+            // tx_by_hash: db.open_tree(b"tx_by_hash").unwrap(),
+            // utxo_by_outpoint: db.open_tree(b"utxo_by_outpoint").unwrap(),
+            // sprout_nullifiers: db.open_tree(b"sprout_nullifiers").unwrap(),
+            // sapling_nullifiers: db.open_tree(b"sapling_nullifiers").unwrap(),
+        }
+    }
+
+    /// Queue a finalized block to be committed to the state.
+    ///
+    /// After queueing a finalized block, this method checks whether the newly
+    /// queued block (and any of its descendants) can be committed to the state.
+    pub fn queue(&mut self, queued_block: QueuedBlock) {
+        let prev_hash = queued_block.block.header.previous_block_hash;
+        self.queued_by_prev_hash.insert(prev_hash, queued_block);
+
+        // Cloning means the closure doesn't hold a borrow of &self,
+        // conflicting with mutable access in the loop below.
+        let hash_by_height = self.hash_by_height.clone();
+        let tip_hash = || {
+            read_tip(&hash_by_height)
+                .expect("inability to look up tip is unrecoverable")
+                .map(|(_height, hash)| hash)
+                .unwrap_or(block::Hash([0; 32]))
+        };
+
+        while let Some(queued_block) = self.queued_by_prev_hash.remove(&tip_hash()) {
+            self.commit_finalized(queued_block)
+        }
+    }
+
+    /// Commit a finalized block to the state.
+    ///
+    /// It's the caller's responsibility to ensure that blocks are committed in
+    /// order. This function is called by [`process_queue`], which does, and is
+    /// intentionally not exposed as part of the public API of the [`SledState`].
```

Comment:
Can we write this sentence with fewer sub-clauses?
```suggestion
    /// order. This function is called by [`process_queue`], which ensures order.
    /// It is intentionally not exposed as part of the public API of the [`SledState`].
```

Reply:
This is now PR #1067.